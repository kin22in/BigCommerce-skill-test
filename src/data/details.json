{
  "title": "Blue Stripe Stoneware Plate",
  "brand": "Kiriko",
  "price": 40,
  "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam at purus pulvinar, placerat turpis ac, interdum metus. In eget massa sed enim hendrerit auctor a eget.",
  "image": "blue-stripe-stoneware-plate.jpg"
}
