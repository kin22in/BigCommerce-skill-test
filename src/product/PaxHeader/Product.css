17 gid=212373379
17 uid=302679766
29 mtime=1528025666.95144565
30 ctime=1528025667.452780002
30 atime=1547173834.024444567
23 SCHILY.dev=16777220
25 SCHILY.ino=8599736705
18 SCHILY.nlink=1
