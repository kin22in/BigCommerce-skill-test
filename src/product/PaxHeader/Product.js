17 gid=212373379
17 uid=302679766
30 mtime=1528025666.963179477
30 ctime=1528025667.459482646
30 atime=1548802939.420882531
23 SCHILY.dev=16777220
25 SCHILY.ino=8599736706
18 SCHILY.nlink=1
