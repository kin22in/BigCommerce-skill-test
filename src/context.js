import React, { Component } from "react";
import details from "./data/details";
import { loadStoreProducts } from "./services/storeServices.js";

const ProductContext = React.createContext();

class ProductProvider extends Component {
  state = {
    products: [],
    cart: [],
    cartTotal: 0,
    cartCount: 0,
    isModalOpen: false,
    modalRef: false,
    productDetail: details
  };
  wrapperRef = "";
  componentDidMount() {
    this.getProductsData();
    window.addEventListener("click", this.dismissModal);
  }

  componentWillUnmount() {
    window.removeEventListener("click", this.dismissModal);
  }

  setWrapperRef = node => {
    this.wrapperRef = node;
  };

  dismissModal = event => {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
      this.setState({ isModalOpen: false });
    }
  };

  openModal = event => {
    event.stopPropagation();
    this.setState({
      isModalOpen: !this.state.isModalOpen
    });
  };

  closeModal = () => {
    this.setState({
      isModalOpen: false
    });
  };

  getProductsData = () => {
    //Calling service to get Data from the Mock Api
    loadStoreProducts().then(res => {
      let products = [];
      res.forEach((item, index) => {
        let singleItem = { ...item };
        //creating a unique if based of index
        singleItem.productId = index + 1;
        // creating new array based of the new object
        products = [...products, singleItem];
      });
      this.setState({
        products
      });
    });
  };

  getItemById = (arr, productId) => {
    return this.state[arr].find(item => item.productId === productId);
  };

  getIndexOf = (arr, productId) => {
    return this.state[arr]
      .map(function(x) {
        return x.productId;
      })
      .indexOf(productId);
  };

  updateCartMetaDetails = tempCart => {
    let getTotal = value => {
      let tempTotal = tempCart.map(item => item[value]);
      return (tempTotal =
        tempTotal.length > 0
          ? tempTotal.reduce((prev, item) => prev + item)
          : 0);
    };

    return { cartTotal: getTotal("total"), cartQuantity: getTotal("quantity") };
  };

  handleDetails = productId => {
    let productDetail = this.getItemById("products", productId);
    this.setState({
      productDetail
    });
  };

  incrementQuanity = (productId, setQuantity) => {
    let tempCart = [...this.state.cart];
    let getExisitngIndex = this.getIndexOf("cart", productId);
    tempCart[getExisitngIndex].quantity =
      setQuantity !== undefined
        ? tempCart[getExisitngIndex].quantity + setQuantity
        : tempCart[getExisitngIndex].quantity + 1;
    tempCart[getExisitngIndex].total =
      tempCart[getExisitngIndex].quantity * tempCart[getExisitngIndex].price;
    let cartMetaDetails = this.updateCartMetaDetails(tempCart);
    this.setState({
      cart: tempCart,
      cartTotal: cartMetaDetails.cartTotal,
      cartCount: cartMetaDetails.cartQuantity
    });
  };

  decrementQuanity = productId => {
    let tempCart = [...this.state.cart];
    let getExisitngIndex = this.getIndexOf("cart", productId);
    if (tempCart[getExisitngIndex].quantity > 0) {
      tempCart[getExisitngIndex].quantity =
        tempCart[getExisitngIndex].quantity - 1;
      tempCart[getExisitngIndex].total =
        tempCart[getExisitngIndex].quantity * tempCart[getExisitngIndex].price;
      let cartMetaDetails = this.updateCartMetaDetails(tempCart);
      this.setState({
        cart: tempCart,
        cartTotal: cartMetaDetails.cartTotal,
        cartCount: cartMetaDetails.cartQuantity
      });
    }
  };

  createNewCartEntry(productId, setQuantity) {
    let productToAdd = this.getItemById("products", productId);
    let tempProduct = { ...productToAdd };
    let tempCart = [...this.state.cart];
    tempProduct.quantity = setQuantity ? setQuantity : 1;
    tempProduct.total = tempProduct.quantity * tempProduct.price;
    tempCart.push(tempProduct);
    let cartMetaDetails = this.updateCartMetaDetails(tempCart);
    this.setState({
      cart: tempCart,
      cartTotal: cartMetaDetails.cartTotal,
      cartCount: cartMetaDetails.cartQuantity
    });
  }

  removeFromCart = (event, productId) => {
    event.stopPropagation();
    let indexOfCartItem = this.getIndexOf("cart", productId);
    let tempCart = [...this.state.cart];
    //remove the item at that index
    tempCart.splice(indexOfCartItem, 1);
    let cartMetaDetails = this.updateCartMetaDetails(tempCart);
    this.setState({
      cart: tempCart,
      cartTotal: cartMetaDetails.cartTotal,
      cartCount: cartMetaDetails.cartQuantity,
    });
  };

  addToCart = (productId, setQuantity) => {
    let indexOfCartItem = this.getIndexOf("cart", productId);
    indexOfCartItem >= 0
      ? this.incrementQuanity(productId, setQuantity)
      : this.createNewCartEntry(productId, setQuantity);
  };

  render() {
    return (
      <ProductContext.Provider
        value={{
          ...this.state,
          removeFromCart: this.removeFromCart,
          openModal: this.openModal,
          closeModal: this.closeModal,
          getIndexOf: this.getIndexOf,
          getItemById: this.getItemById,
          setWrapperRef: this.setWrapperRef,
          incrementQuanity: this.incrementQuanity,
          decrementQuanity: this.decrementQuanity,
          handleDetails: this.handleDetails,
          addToCart: this.addToCart
        }}
      >
        {this.props.children}
      </ProductContext.Provider>
    );
  }
}

const ProductConsumer = ProductContext.Consumer;

export { ProductProvider, ProductConsumer };
