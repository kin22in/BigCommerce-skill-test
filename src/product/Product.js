import React, { Component } from "react";
import "./Product.css";
import { ProductConsumer } from "../context.js";
class Product extends Component {
  constructor(props) {
    super(props);
    this.state = {
      id: props.match.params.id
    };
  }

  render() {
    return (
      <ProductConsumer>
        {({
          productDetail,
          cart,
          incrementQuanity,
          getIndexOf,
          getItemById,
          addToCart,
          decrementQuanity
        }) => {
          const {
            title,
            brand,
            price,
            image,
            productId,
            description
          } = productDetail;
          return (
            <div className="container">
              <div className="row text-center">
                <div className="col-12 mt-3 mb-5">HOME / PLATES /{title}</div>
                <div className="col-8">
                  <img
                    className="img-fluid"
                    src={`/media/${image}`}
                    alt={title}
                  />
                </div>
                <div className="col-4 px-3">
                  <p className="text-muted">{brand}</p>
                  <p className="h1">{title}</p>
                  <p className="text-muted">${price}</p>
                  <p className="text-muted">{description}</p>
                  <hr className="px-1" />
                  <div className="d-flex justify-content-center align-items-stretch">
                    <div className="bg-light border mx-1 px-3 d-flex align-items-center">
                      {getIndexOf("cart", productId) >= 0
                        ? getItemById("cart", productId).quantity
                        : 0}
                    </div>
                    <div className="mr-3 d-flex flex-column">
                      <div className="">
                        <button
                          onClick={() => {
                            addToCart(productId);
                          }}
                          className="w-100 btn border mb-1 btn-light"
                        >
                          +
                        </button>
                      </div>
                      <div className="">
                        <button
                          onClick={() => {
                            decrementQuanity(productId);
                          }}
                          className="btn w-100 border btn-light"
                        >
                          -
                        </button>
                      </div>
                    </div>
                    <button
                      onClick={() => {
                        addToCart(productId);
                      }}
                      className="rounded-0 btn btn-dark border"
                    >
                      ADD TO CART
                    </button>
                  </div>
                </div>
              </div>
            </div>
          );
        }}
      </ProductConsumer>
    );
  }
}

export default Product;
