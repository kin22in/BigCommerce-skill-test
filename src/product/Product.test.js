import React from "react";
import {
  render,
  fireEvent,
  cleanup,
  getAllByTestId,
  getByTestId
} from "react-testing-library";
import { BrowserRouter } from "react-router-dom";
import details from "../data/details";
import { Link, Route, Router, Switch } from "react-router-dom";
import Product from "../product/Product.js";
import { ProductContext } from "../context.js";

let detail = {
  title: "Blue Stripe Stoneware Plate",
  brand: "Kiriko",
  price: 40,
  productId: 1,
  description:
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam at purus pulvinar, placerat turpis ac, interdum metus. In eget massa sed enim hendrerit auctor a eget.",
  image: "blue-stripe-stoneware-plate.jpg"
};
afterEach(cleanup);

test("Increment quantity on clicking +", () => {
  const div = document.createElement("div");
  const { getByText, getByTestId, debug } = render(
    <ProductContext.Provider value={{ productDetail: detail }}>
      <BrowserRouter>
        <Product />
      </BrowserRouter>
    </ProductContext.Provider>
  );

  const increment = getByText("+");
  const quantityCount = getByTestId("quantityCount");
  expect(quantityCount.textContent).toEqual("1");
  fireEvent.click(increment);
  expect(quantityCount.textContent).toEqual("2");
});

test("Decrement quantity on clicking -", () => {
  const div = document.createElement("div");
  const { getByText, getByTestId, debug } = render(
    <ProductContext.Provider value={{ productDetail: detail }}>
      <BrowserRouter>
        <Product />
      </BrowserRouter>
    </ProductContext.Provider>
  );

  const decrement = getByText("-");
  const quantityCount = getByTestId("quantityCount");
  expect(quantityCount.textContent).toEqual("1");
  fireEvent.click(decrement);
  expect(quantityCount.textContent).toEqual("0");
});

test("Creating Snapshot", () => {
  const div = document.createElement("div");
  const { container } = render(
    <ProductContext.Provider value={{ productDetail: detail }}>
      <BrowserRouter>
        <Product />
      </BrowserRouter>
    </ProductContext.Provider>
  );

  expect(container.firstChild).toMatchSnapshot();
});
