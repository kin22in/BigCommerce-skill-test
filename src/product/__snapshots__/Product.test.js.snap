// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creating Snapshot 1`] = `
<div
  class="container"
>
  <div
    class="row text-center"
  >
    <div
      class="col-12 mt-3 mb-5"
    >
      HOME / PLATES /
      Blue Stripe Stoneware Plate
    </div>
    <div
      class="col-8"
    >
      <img
        class="img-fluid"
        src="/media/blue-stripe-stoneware-plate.jpg"
        alt="Blue Stripe Stoneware Plate"
      />
    </div>
    <div
      class="col-4 px-3"
    >
      <p
        class="text-muted"
      >
        Kiriko
      </p>
      <p
        class="h1"
      >
        Blue Stripe Stoneware Plate
      </p>
      <p
        class="text-muted"
      >
        $
        40
      </p>
      <p
        class="text-muted"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam at purus pulvinar, placerat turpis ac, interdum metus. In eget massa sed enim hendrerit auctor a eget.
      </p>
      <hr
        class="px-1"
      />
      <div
        class="d-flex justify-content-center align-items-stretch"
      >
        <div
          data-testid="quantityCount"
          class="bg-light border mx-1 px-3 d-flex align-items-center"
        >
          1
        </div>
        <div
          class="mr-3 d-flex flex-column"
        >
          <div
            class=""
          >
            <button
              class="w-100 btn border mb-1 btn-light"
            >
              +
            </button>
          </div>
          <div
            class=""
          >
            <button
              class="btn w-100 border btn-light"
            >
              -
            </button>
          </div>
        </div>
        <button
          class="rounded-0 btn btn-dark border"
        >
          ADD TO CART
        </button>
      </div>
    </div>
  </div>
</div>
`;
